src/cac/vsss.rs

```diff
@@ -0,0 +1,181 @@
+use bitcoin::key::rand;
+use eyre::{Result, eyre};
+use k256::elliptic_curve::PrimeField;
+use k256::{ProjectivePoint, Scalar};
+use std::ops::Add;
+use std::ops::Mul;
+
+// we use this for both polynomials over scalars and over projective points
+pub struct Polynomial<T>(Vec<T>);
+
+impl<T> Polynomial<T>
+where
+    T: Add<T, Output = T> + Mul<Scalar, Output = T> + Copy,
+{
+    // todo: max x an int
+    fn eval_at(&self, x: Scalar) -> T {
+        self.0
+            .iter()
+            .skip(1)
+            .fold((self.0[0], Scalar::ONE), |(acc, prev_factor), &elem| {
+                let new_factor = prev_factor * x;
+                (acc + elem * new_factor, new_factor)
+            })
+            .0
+    }
+}
+
+impl Polynomial<Scalar> {
+    pub fn rand(degree: usize) -> Self {
+        Self(
+            (0..degree + 1)
+                .map(|_| Scalar::generate_vartime(&mut rand::thread_rng()))
+                .collect(),
+        )
+    }
+    pub fn coefficient_commits(&self) -> PolynomialCommits {
+        PolynomialCommits(Polynomial(
+            self.0
+                .iter()
+                .map(|x| ProjectivePoint::GENERATOR * x)
+                .collect(),
+        ))
+    }
+    // shares are return with 0-based index. However, we evaluate share i at
+    // x = i+1, since the value at x=0 represents the secret
```

Comment:
Could we evaluate share i at x=i, or would that break the security? Having `x=0` represent the secret is only for the output label, and there we could just not send share[0]